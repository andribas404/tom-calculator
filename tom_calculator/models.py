"""Models."""
import logging
from typing import Type

import sqlalchemy as sa
from sqlalchemy.dialects.postgresql import UUID

from tom_calculator.database import Base

logger = logging.getLogger(__name__)

TBase = Type[Base]


class Tax(Base):
    """Tax model."""
    __tablename__ = 'taxes'
    __table_args__ = {'comment': 'Table with tax rates per state.'}

    id = sa.Column(
        sa.Integer,
        primary_key=True,
        doc='Tax identifier.',
        comment='Tax identifier.',
    )
    state_name = sa.Column(
        sa.String(20),
        nullable=False,
        index=True,
        unique=True,
        doc='Name of the state.',
        comment='Name of the state.',
    )
    rate = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Rate for the current state measured in `%`.',
        comment='Rate for the current state measured in `%`.',
    )


class Discount(Base):
    """Discount model."""
    __tablename__ = 'discounts'
    __table_args__ = {'comment': 'Table with discount rates applied to the amount.'}

    id = sa.Column(
        sa.Integer,
        primary_key=True,
        doc='Dicsount identifier.',
        comment='Dicsount identifier.',
    )
    amount = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        index=True,
        doc='Amount for which this discount is applicable.',
        comment='Amount for which this discount is applicable.',
    )
    rate = sa.Column(
        sa.Numeric(16, 0),
        nullable=False,
        doc='Discount rate applied to amount measured in `%`.',
        comment='Discount rate applied to amount measured in `%`.',
    )


class Order(Base):
    """Order model."""
    __tablename__ = 'orders'
    __table_args__ = {'comment': 'Table with orders.'}
    __mapper_args__ = {'eager_defaults': True}

    id = sa.Column(  # type: ignore[misc]
        UUID,
        primary_key=True,
        server_default=sa.func.uuid_generate_v4(),
        doc='Order identifier.',
        comment='Order identifier.',
    )
    ts = sa.Column(  # type: ignore[misc]
        sa.DateTime,
        server_default=sa.func.now(),
        nullable=False,
        index=True,
        doc='Order timestamp.',
        comment='Order timestamp.',
    )
    amount = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Amount of order received from calculator.',
        comment='Amount of order received from calculator.',
    )
    # after_discount = amount - discount
    after_discount = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Amount of order after discount has been applied.',
        comment='Amount of order after discount has been applied.',
    )
    # tax = after_discount * rate
    tax = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Tax sum calculated from discounted amount.',
        comment='Tax sum calculated from discounted amount.',
    )
    # total = after_discount + tax
    total = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Total amount of order after discount and tax.',
        comment='Total amount of order after discount and tax.',
    )
