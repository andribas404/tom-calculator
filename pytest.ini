[pytest]
addopts =
    --cov tom_calculator --cov-report term-missing
    --verbose
    --color=yes
    --show-capture=no

markers =
    unit: Unit tests
    integration: Integration tests
