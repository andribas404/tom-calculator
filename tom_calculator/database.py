import logging
from asyncio import current_task
from contextlib import AbstractAsyncContextManager, asynccontextmanager
from typing import Callable

from sqlalchemy import orm
from sqlalchemy.ext.asyncio import AsyncSession, async_scoped_session, create_async_engine

logger = logging.getLogger(__name__)

Base = orm.declarative_base()

TSession = Callable[..., AbstractAsyncContextManager[orm.Session]]


class Database:
    """Database."""
    def __init__(self, db_dsn: str) -> None:
        self._engine = create_async_engine(db_dsn, future=True, echo=True)
        self._async_session_factory = orm.sessionmaker(
            class_=AsyncSession,
            autocommit=False,
            autoflush=False,
            expire_on_commit=False,
            bind=self._engine,
        )
        self._session = async_scoped_session(self._async_session_factory, scopefunc=current_task)

    async def create_database(self) -> None:
        async with self._engine.begin() as conn:
            await conn.run_sync(Base.metadata.drop_all)
            await conn.run_sync(Base.metadata.create_all)

    async def close(self) -> None:
        await self._engine.dispose()

    @asynccontextmanager
    async def session(self) -> TSession:
        async with self._session() as session:
            try:
                yield session
                await session.commit()
            except Exception:
                logger.exception('Session rollback because of exception')
                await session.rollback()
                raise
            finally:
                await session.close()
