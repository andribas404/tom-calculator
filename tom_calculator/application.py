import logging
from fastapi import FastAPI

from tom_calculator.containers import Container
from tom_calculator.util import get_config_path
from tom_calculator import endpoints

logger = logging.getLogger(__name__)


def create_app() -> FastAPI:
    config_path = get_config_path()
    container = Container()
    container.config.from_yaml(config_path)
    container.wire(modules=[endpoints])

    app = FastAPI()
    app.container = container
    app.include_router(endpoints.router)

    return app


app = create_app()
