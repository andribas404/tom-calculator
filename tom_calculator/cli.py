import asyncio
import os
import subprocess

from dependency_injector.wiring import inject, Provide
import typer

from tom_calculator import services
from tom_calculator.application import create_app

app = typer.Typer()


@inject
def load(
    datadir: str,
    loader_service: services.LoaderService = Provide['loader_service'],
):
    asyncio.run(loader_service.load(datadir))


@app.callback()
def main(
    ctx: typer.Context,
):
    main_app = create_app()
    ctx.obj = main_app


@app.command()
def migrate():
    typer.echo('Starting migration...')
    subprocess.run(['alembic', 'migrate', 'head'])


@app.command()
def migrate_data(
):
    typer.echo('Migrating data...')
    datadir = os.getenv('TOM_DATA')
    load(datadir)


if __name__ == '__main__':
    app()
