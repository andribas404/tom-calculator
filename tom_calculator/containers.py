from dependency_injector import containers, providers

from tom_calculator.database import Database
from tom_calculator.services import OrderService


class Container(containers.DeclarativeContainer):

    config = providers.Configuration()

    db = providers.Singleton(Database, db_dsn=config.db.async_dsn)

    order_service = providers.Factory(
        OrderService,
        session_factory=db.provided.session,
    )
