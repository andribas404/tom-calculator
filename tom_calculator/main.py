"""Main module.

Contains application instance.
"""
import logging

import uvicorn

from tom_calculator.application import create_app

logger = logging.getLogger(__name__)

app = create_app()


if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8000)
