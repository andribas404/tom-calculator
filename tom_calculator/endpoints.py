"""API endpoints."""
import logging
from typing import Any
import uuid

from dependency_injector.wiring import Provide, inject
from fastapi import APIRouter, Depends, HTTPException, Response, status
from starlette.responses import RedirectResponse

from tom_calculator import schemas, services
from tom_calculator.containers import Container

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get('/')
async def root() -> Any:
    """Root endpoint.

    Redirects to openapi.
    """
    response = RedirectResponse(url='/docs#/default/order_create_order_post')
    return response


@router.post(
    '/order',
    status_code=status.HTTP_201_CREATED,
    response_model=schemas.OrderItemOut,
)
@inject
async def order_create(
    item: schemas.CalculatorIn,
    order_service: services.OrderService = Depends(Provide[Container.order_service]),
) -> Any:
    """Endpoint for creating orders from calculator input.

    Returns created order.
    """
    try:
        return await order_service.create(item)
    except services.TaxNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f'Tax for state {item.state_name} not exist.'
        )


@router.get(
    '/order/{item_id}',
    response_model=schemas.OrderItemOut,
)
@inject
async def order_get(
    item_id: uuid.UUID,
    order_service: services.OrderService = Depends(Provide[Container.order_service]),
) -> Any:
    """Endpoint for retrieving order by id."""
    try:
        return await order_service.get(item_id)
    except services.OrderNotFoundError:
        return Response(status_code=status.HTTP_404_NOT_FOUND)
