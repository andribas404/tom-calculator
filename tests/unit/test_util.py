"""Test util.py."""
from decimal import Decimal
from pathlib import Path
from unittest import mock

import pytest

from tom_calculator import util


class TestGetConfigPath:
    """Test get_config_path."""
    @mock.patch('tom_calculator.util.os.getenv')
    def test_get_config_path(self, mock_getenv):
        """Test get_config_path."""
        mock_getenv.return_value = 'config'
        result = util.get_config_path()
        assert result == Path('config')


class TestGetDatadir:
    """Test get_datadir."""
    @mock.patch('tom_calculator.util.os.getenv')
    def test_get_config_path(self, mock_getenv):
        """Test get_datadir."""
        mock_getenv.return_value = 'config'
        result = util.get_datadir()
        assert result == Path('config')


class TestLoadCsv:
    """Test load_csv."""
    def test_load_csv_discounts(self, data_discounts, snapshot):
        """Test load_csv."""
        result = util.load_csv(data_discounts)
        assert result == snapshot

    def test_load_csv_taxes(self, data_taxes, snapshot):
        """Test load_csv."""
        result = util.load_csv(data_taxes)
        assert result == snapshot


class TestRoundUp:
    """Test round_up."""
    @pytest.mark.parametrize(('amount', 'expected'), [
        (Decimal('0'), Decimal('0.')),
        (Decimal('23'), Decimal('23')),
        (Decimal('23.05'), Decimal('23.05')),
        (Decimal('0.001'), Decimal('0.01')),
        (Decimal('0.234'), Decimal('0.24')),
        (Decimal('0.22999999'), Decimal('0.23')),
        (Decimal('-0.00001'), Decimal('0.')),
        (Decimal('232.05555'), Decimal('232.06')),
    ])
    def test_round_up(self, amount, expected):
        """Test ."""
        result = util.round_up(amount)
        assert result == expected


class TestRoundDown:
    """Test round_down."""
    @pytest.mark.parametrize(('amount', 'expected'), [
        (Decimal('0'), Decimal('0.')),
        (Decimal('23'), Decimal('23')),
        (Decimal('23.05'), Decimal('23.05')),
        (Decimal('0.001'), Decimal('0.')),
        (Decimal('0.234'), Decimal('0.23')),
        (Decimal('0.22999999'), Decimal('0.22')),
        (Decimal('-0.00001'), Decimal('-0.01')),
        (Decimal('232.05555'), Decimal('232.05')),
    ])
    def test_round_down(self, amount, expected):
        """Test round_down."""
        result = util.round_down(amount)
        assert result == expected
