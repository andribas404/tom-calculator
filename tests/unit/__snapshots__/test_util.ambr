# name: TestLoadCsv.test_load_csv_discounts
  <class 'list'> [
    <class 'dict'> {
      'amount': '1000',
      'rate': '3',
    },
    <class 'dict'> {
      'amount': '5000',
      'rate': '5',
    },
    <class 'dict'> {
      'amount': '7000',
      'rate': '7',
    },
    <class 'dict'> {
      'amount': '10000',
      'rate': '10',
    },
    <class 'dict'> {
      'amount': '50000',
      'rate': '15',
    },
  ]
---
# name: TestLoadCsv.test_load_csv_taxes
  <class 'list'> [
    <class 'dict'> {
      'rate': '6.85',
      'state_name': 'UT',
    },
    <class 'dict'> {
      'rate': '8',
      'state_name': 'NV',
    },
    <class 'dict'> {
      'rate': '6.25',
      'state_name': 'TX',
    },
    <class 'dict'> {
      'rate': '4',
      'state_name': 'AL',
    },
    <class 'dict'> {
      'rate': '8.25',
      'state_name': 'CA',
    },
  ]
---
